---
title: "scrna_pipe"
author: "Zhe"
date: "June 8, 2017"
output: html_document
---


```{r}
# scRNA-seq pipeline
# Keep track of In vitro transcription RNA molecules
# Keep track of final PCR amplification products

# Input: demultiplexed SAM file
# Output: count matrix after UMI correction, 


# required packages:
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(gtools))
suppressPackageStartupMessages(library(stringdist))

```


```{r}
##  parameters


# umi.edit = as.numeric(opt$options$umi_edit_dist)
# umi.window = as.numeric(opt$options$window)
# umi.length = as.numeric(opt$options$umi_length)
test.sam <- "../data/CS_1017_sample_0001.sam"
fname <- last(strsplit(substr(test.sam, start=1,
                              stop=nchar(test.sam)-4), split="/", fixed=T)[[1]])

# acceptable umi sequence mismatches 
umi.edit <- 1

# default sliding window for fragmentation discrepencies
umi.max.gap <- 20

# default sliding window for alignment position correction
pos.max.gap <- 3

# 
umi.length <- 5
duplication.offset = 1024L  # SAM specification for duplication in FLAG column

output.dir <- "../res/"

# SAM file directory
out.sam <- paste0(output.dir, fname, ".tab")

umi.stats <- paste0(output.dir, fname, "_UMI_stats.tab")

umi.qc <- paste0(output.dir, fname, "_UMI_QC_plots.pdf")

umi.original.counts <- paste0(output.dir, fname, "_UMI_original_counts.txt")

umi.inferred.counts <- paste0(output.dir, fname, "_UMI_inferred_counts.txt")

umi.position.counts <- paste0(output.dir, fname, "_UMI_position_counts.txt")
  
# remove output file if exists
#if (file.exists(out.sam)) file.remove(out.sam)

umi.matrix <- c()


```


```{r}
# functions

read.sam <- function(samfile){
  # get the maximum number of columns per row
  maxncol <- max(count.fields(samfile, sep="\t", quote="", comment.char=""))
  
  # read in SAM file
  sam <- read.table(samfile, 
              sep="\t",
              quote="",
              fill=T, 
              header=F,
              stringsAsFactors=F,
              na.strings=NULL,
              comment.char="@",
              col.names=1:maxncol)
  
  colnames(sam)[1:11] <- c("qname", "flag", "rname", "position", "mapq", "cigar",
                             "rnext", "pnext", "tlen", "seq", "qual")
  
  # convert to data.table object
  samdt <- data.table(sam, check.names=T)
  return (samdt)
}

umi.mismatch.correction <- function(samdt, current.ref, umi.max.gap, umi.edit) {
  cat("Now processing", current.ref, "...\n")
  
  # Add inferred_umi info to reference data table
  rdt <- samdt[rname == current.ref,]
  rdt[, c("umi", "inferred_umi") := 
      tstrsplit(qname, ":", fixed=TRUE, 
                keep=length(tstrsplit(qname, ":", fixed=TRUE)))]
  
  # Correct UMIs with sequencing errors by looking at UMIs in surrounding region
  
  # get all alignment positions
  unique.pos <- sort(unique(rdt$position))
  
  unique.pos.list <- get.adj.pos.list(unique.pos, umi.max.gap)
  
  # for each IVT fragment
  for (i in unique.pos.list) {
    all.umi.count <- sort(table(c(rdt[position %in% i, umi])))
    
    if (length(all.umi.count) > 1) {
      # temporary solution with only one iteration
      # need a recursive solution for some special cases
      sdm <- stringdistmatrix(names(all.umi.count), names(all.umi.count))
	    diag(sdm) <- 100
	    rownames(sdm) <- names(all.umi.count)
		  colnames(sdm) <- names(all.umi.count)
      
      for (j in colnames(sdm)) {
        if (min(sdm[j,]) <= umi.edit) {
          sdm.edit.ind <- max(which(sdm[j,] <= umi.edit))
          # correct current umi j within position group i
          if (which(rownames(sdm) == j) < sdm.edit.ind) {
            rdt[umi == j & position %in% i, inferred_umi := colnames(sdm)[sdm.edit.ind]]
          }
        }
      }
    }
  }
  return (rdt)
}


get.adjacent.unique.pos <- function(unique.pos, gap) {
  adjs <- diff(unique.pos) <= gap
  ind <- which(adjs == T)
  plusone <- ind + 1
  return (unique.pos[sort(union(ind, plusone))])
}


get.adj.pos.list <- function(unique.pos, gap){
  adjs <- diff(unique.pos) <= gap
  n <- sum(adjs == F) + 1
  res <- vector("list", n) 
  
  i <- 1 # Alignment position group
  j <- 1 # index of position
  while (i <= n) {
    if (j > length(adjs)) {
      res[[i]] <- c(res[[i]], unique.pos[j])
      i <- i + 1
    } else if (adjs[j] == T) {
      res[[i]] <- c(res[[i]], unique.pos[j])
    } else if (adjs[j] == F) {
      res[[i]] <- c(res[[i]], unique.pos[j])
      i <- i + 1
    } 
    j <- j + 1
  }
  return (res)
}


get.position.with.most.reads <- function(rdt.sub) {
  res <- sort(table(rdt.sub[,position]), decreasing = T)
  return (sort(as.numeric(names(res[which(res == res[1])])))[1])
}


alignment.position.correction <- function(rdt, pos.max.gap) {
  rdt[,inferred_pos := position]
  unique.pos <- sort(unique(rdt$position))
  
  # consider only adjacent positions with gap <= pos.max.gap
  adj.unique.pos <- get.adjacent.unique.pos(unique.pos, pos.max.gap)
  pos.group <- get.adj.pos.list(adj.unique.pos, pos.max.gap)
  # correct alignment position error
  for (i in pos.group) {
    rdt.sub <- rdt[position %in% i, ]
    unique.umi.count <- table(rdt.sub$inferred_umi)

    for (j in 1:length(unique.umi.count)) { 
      rdt.sub.sub <- rdt.sub[inferred_umi == names(unique.umi.count)[j], ]
      rdt[position %in% i & inferred_umi == names(unique.umi.count)[j], 
          inferred_pos := get.position.with.most.reads(rdt.sub.sub)]
    }
  }
  return (rdt)
}


get.pcr.duplicates <- function(rdt) {
  reads <- rdt[,.(inferred_umi, inferred_pos, rname)]
  unique.fragments <- unique(reads[order(reads$inferred_pos)])
  for (i in 1:nrow(unique.fragments)) {
    frag <- unique.fragments[i,]
    n <- nrow(rdt[inferred_umi==frag[[1]] & inferred_pos==frag[[2]],])
    unique.fragments[i, num := n]
  }
  return (unique.fragments)
}

```








```{r}

# read in SAM file
samdt <- read.sam(test.sam)

# get reference sequence names
chr <- mixedsort(setdiff(unique(samdt[,rname]), "*"))
res.table <- list()
num.pcr.products <- list()


# for each reference sequence name (chr)
for (current.ref in chr) {
  # correct umi mismatch
  rdt <- umi.mismatch.correction(samdt, current.ref, umi.max.gap, umi.edit)
  # correct alignment position error
  rdt <- alignment.position.correction(rdt, pos.max.gap)
  
  
  # Now that UMIs have been fixed, go through each position and pick one read to
	# represent each UMI. Positions with more reads with that UMI and that are more 5'
	# are given higher priority. One read randomly selected to be the non-duplicate
  
  # Metrics to report:
  # Number of PCR priducts per IVT fragment
  num.pcr.products <- rbindlist(list(num.pcr.products, get.pcr.duplicates(rdt)),
                                use.names=F, fill=F, idcol=F)
  
  # Distribution of average UMI edit distance of all reads


	res.table <- rbindlist(list(res.table, 
	                   rdt[,c("rname", "position","inferred_pos",
	                          "umi", "inferred_umi")]),
	                   use.names=F, fill=F, idcol=F)
}

# Number of reads with mismatches in UMI
  num.umi.mismatch <- nrow(res.table[umi != inferred_umi,])

# Percentage of reads with shifts in alignment position
  num.pos.shift <- nrow(res.table[position != inferred_pos,])


stats.label <- c("Number of aligned reads:", 
                "Number of duplicates:", 
                "Percent duplication:",
                "Number of unique UMIs:",
                "Number of unique UMIs after correction:",
                "Number of reads with mismatches in UMI:",
                "Percentage of reads with mismatches in UMI:")

stats <- c(nrow(umi.matrix), 
          sum(umi.matrix$duplicate, na.rm = T),
          sum(umi.matrix$duplicate, na.rm = T)/nrow(umi.matrix),
          length(unique(umi.matrix[,umi])), 
          length(unique(umi.matrix[,inferred_umi])),
          sum(umi.matrix[,umi] != umi.matrix[,inferred_umi]),
          sum(umi.matrix[,umi] != umi.matrix[,inferred_umi]) / nrow(umi.matrix))

fwrite(data.table(stats.label, stats), umi.stats, sep="\t",
            quote=FALSE, row.names=FALSE, col.names=FALSE)

# Calculate relative abundance of UMIs
umi.table <- table(umi.matrix[,umi])
umi.inferred.table <- table(umi.matrix[,inferred_umi])

# Base percentage of UMIs at each position
base.matrix <- c()
for(i in 1:umi.length) {
  base.matrix <- cbind(base.matrix, table(substring(unique(umi.matrix[,umi]), i, i)))
}

#base.matrix.frac = sweep(base.matrix, 2, length(unique(umi.matrix[,umi])), "/")
base.matrix.frac <- base.matrix / length(unique(umi.matrix[,umi]))

base.inferred.matrix <- c()
for(i in 1:umi.length) {
  base.inferred.matrix <- cbind(base.inferred.matrix, 
                               table(substring(unique(umi.matrix[,inferred_umi]), i, i)))
}

#base.inferred.matrix.frac = 
#  sweep(base.inferred.matrix, 2, length(unique(umi.matrix[,4])), "/")
base.inferred.matrix.frac <- base.inferred.matrix / length(unique(umi.matrix[,inferred_umi]))


pdf(umi.qc, useDingbats=FALSE)

hist(log2(umi.table), freq=FALSE, xlab="Log2(number of reads per UMI)", ylab="UMI Fraction",
     lwd=2,breaks=20, main="Original UMIs")

hist(log2(umi.inferred.table), freq=FALSE, xlab="Log2(number of reads per UMI)", 
     ylab="UMI Fraction", lwd=2, breaks=20, main="Inferred UMIs")

barplot(base.matrix.frac, col=1:4, xlab="UMI position", names.arg=(1:umi.length),
        ylab="Percent")
        
barplot(base.inferred.matrix.frac, col=1:4, xlab="UMI position", names.arg=(1:umi.length),
        ylab="Percent")

plot(1:10, type="n")
legend("center", rownames(base.matrix.frac), col=1:4, pch=15)
dev.off()

o <- order(umi.table, decreasing=TRUE)

fwrite(data.table(umi.table)[o,],
       umi.original.counts,
       sep="\t", quote=FALSE, row.names=FALSE, col.names=FALSE)

o <- order(umi.inferred.table, decreasing=TRUE)

fwrite(data.table(umi.inferred.table)[o,],
       umi.inferred.counts,
       sep="\t", quote=FALSE, row.names=FALSE, col.names=FALSE)

umi.loc.table <- table(umi.matrix[,paste(c(rname,position,inferred_umi),collapse="_"),
                                  by = 1:nrow(umi.matrix)][,V1])

o <- order(umi.loc.table, decreasing=TRUE)

fwrite(data.table(umi.loc.table)[o,], 
       umi.position.counts, 
       sep="\t", quote=FALSE, row.names=FALSE, col.names=FALSE)


```


