---
title: "scrna_pipe"
author: "Zhe"
date: "June 8, 2017"
output: html_document
---


```{r}
# scRNA-seq pipeline
# Keep track of In vitro transcription RNA molecules
# Keep track of final PCR amplification products

# Input: demultiplexed SAM file
# Output: count matrix after UMI correction, 


# required packages:
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(gtools))
suppressPackageStartupMessages(library(stringdist))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(ggthemes))
suppressPackageStartupMessages(library(scales))

```


```{r}
##  parameters


# umi.edit = as.numeric(opt$options$umi_edit_dist)
# umi.window = as.numeric(opt$options$window)
# umi.length = as.numeric(opt$options$umi_length)
test.sam <- "../data/CS_1017_sample_0001.sam"
fname <- last(strsplit(substr(test.sam, start=1,
                              stop=nchar(test.sam)-4), split="/", fixed=T)[[1]])

# acceptable umi sequence mismatches 
umi.edit <- 1

# default sliding window for fragmentation discrepencies
umi.max.gap <- 20

# default sliding window for alignment position correction
pos.max.gap <- 3

# 
umi.length <- 5
duplication.offset = 1024L  # SAM specification for duplication in FLAG column

output.dir <- "../res/"

# SAM file directory
out.sam <- paste0(output.dir, fname, ".tab")

umi.stats <- paste0(output.dir, fname, "_UMI_stats.tab")

umi.qc <- paste0(output.dir, fname, "_UMI_QC_plots.pdf")

umi.original.counts <- paste0(output.dir, fname, "_UMI_original_counts.txt")

umi.inferred.counts <- paste0(output.dir, fname, "_UMI_inferred_counts.txt")

umi.position.counts <- paste0(output.dir, fname, "_UMI_position_counts.txt")
  
# remove output file if exists
#if (file.exists(out.sam)) file.remove(out.sam)

umi.matrix <- c()


```


```{r}
# ggplot publication theme
# adapted from Koundinya Desiraju
# https://rpubs.com/Koundy/71792

theme_Publication <- function(base_size=12, base_family="sans") {
      (ggthemes::theme_foundation(base_size=base_size, base_family=base_family)
       + ggplot2::theme(plot.title = element_text(face = "bold",
                                         size = rel(1), hjust = 0.5),
               text = element_text(),
               panel.background = element_rect(colour = NA),
               plot.background = element_rect(colour = NA),
               panel.border = element_rect(colour = NA),
               axis.title = element_text(face = "bold",size = rel(1)),
               axis.title.y = element_text(angle=90,vjust =2),
               axis.title.x = element_text(vjust = -0.2),
               axis.text = element_text(), 
               axis.line = element_line(colour="black"),
               axis.ticks = element_line(),
               panel.grid.major = element_line(colour="#f0f0f0"),
               panel.grid.minor = element_blank(),
               legend.key = element_rect(colour = NA),
               legend.position = "right",
               legend.direction = "vertical",
               legend.key.size= unit(0.5, "cm"),
               legend.margin = margin(0),
               legend.title = element_text(face="bold"),
               plot.margin=unit(c(10,5,5,5),"mm"),
               strip.background=element_rect(colour="#f0f0f0",fill="#f0f0f0"),
               strip.text = element_text(face="bold")
          ))
      
}

cpalette <- c("#386cb0","#fdb462","#7fc97f","#ef3b2c","#662506","#0A3708",
                "#a6cee3","#fb9a99","#984ea3","#ffff33","#000000", "#756682")

scale_fill_Publication <- function(...){
      ggplot2::discrete_scale("fill","Publication",manual_pal(values = cpalette), ...)

}

scale_colour_Publication <- function(...){
      ggplot2::discrete_scale("colour","Publication",manual_pal(values = cpalette), ...)

}
```




```{r}
# functions

read.sam <- function(samfile){
  # get the maximum number of columns per row
  maxncol <- max(count.fields(samfile, sep="\t", quote="", comment.char=""))
  
  # read in SAM file
  sam <- read.table(samfile, 
              sep="\t",
              quote="",
              fill=T, 
              header=F,
              stringsAsFactors=F,
              na.strings=NULL,
              comment.char="@",
              col.names=1:maxncol)
  
  colnames(sam)[1:11] <- c("qname", "flag", "rname", "position", "mapq", "cigar",
                             "rnext", "pnext", "tlen", "seq", "qual")
  
  # convert to data.table object
  samdt <- data.table(sam, check.names=T)
  return (samdt)
}

umi.mismatch.correction <- function(samdt, current.ref, umi.max.gap, umi.edit) {
  cat("Now processing", current.ref, "...\n")
  
  # Add inferred_umi info to reference data table
  rdt <- samdt[rname == current.ref,]
  rdt[, c("umi", "inferred_umi") := 
      tstrsplit(qname, ":", fixed=TRUE, 
                keep=length(tstrsplit(qname, ":", fixed=TRUE)))]
  
  # Correct UMIs with sequencing errors by looking at UMIs in surrounding region
  
  # get all alignment positions
  unique.pos <- sort(unique(rdt$position))
  
  unique.pos.list <- get.adj.pos.list(unique.pos, umi.max.gap)
  
  # for each IVT fragment
  for (i in unique.pos.list) {
    all.umi.count <- sort(table(c(rdt[position %in% i, umi])))
    
    if (length(all.umi.count) > 1) {
      # temporary solution with only one iteration
      # need a recursive solution for some special cases
      sdm <- stringdistmatrix(names(all.umi.count), names(all.umi.count))
	    diag(sdm) <- 100
	    rownames(sdm) <- names(all.umi.count)
		  colnames(sdm) <- names(all.umi.count)
      
      for (j in colnames(sdm)) {
        if (min(sdm[j,]) <= umi.edit) {
          sdm.edit.ind <- max(which(sdm[j,] <= umi.edit))
          # correct current umi j within position group i
          if (which(rownames(sdm) == j) < sdm.edit.ind) {
            rdt[umi == j & position %in% i, inferred_umi := colnames(sdm)[sdm.edit.ind]]
          }
        }
      }
    }
  }
  return (rdt)
}


get.adjacent.unique.pos <- function(unique.pos, gap) {
  adjs <- diff(unique.pos) <= gap
  ind <- which(adjs == T)
  plusone <- ind + 1
  return (unique.pos[sort(union(ind, plusone))])
}


get.adj.pos.list <- function(unique.pos, gap){
  adjs <- diff(unique.pos) <= gap
  n <- sum(adjs == F) + 1
  res <- vector("list", n) 
  
  i <- 1 # Alignment position group
  j <- 1 # index of position
  while (i <= n) {
    if (j > length(adjs)) {
      res[[i]] <- c(res[[i]], unique.pos[j])
      i <- i + 1
    } else if (adjs[j] == T) {
      res[[i]] <- c(res[[i]], unique.pos[j])
    } else if (adjs[j] == F) {
      res[[i]] <- c(res[[i]], unique.pos[j])
      i <- i + 1
    } 
    j <- j + 1
  }
  return (res)
}


get.position.with.most.reads <- function(rdt.sub) {
  res <- sort(table(rdt.sub[,position]), decreasing = T)
  return (sort(as.numeric(names(res[which(res == res[1])])))[1])
}


alignment.position.correction <- function(rdt, pos.max.gap) {
  rdt[,inferred_pos := position]
  unique.pos <- sort(unique(rdt$position))
  
  # consider only adjacent positions with gap <= pos.max.gap
  adj.unique.pos <- get.adjacent.unique.pos(unique.pos, pos.max.gap)
  pos.group <- get.adj.pos.list(adj.unique.pos, pos.max.gap)
  # correct alignment position error
  for (i in pos.group) {
    rdt.sub <- rdt[position %in% i, ]
    unique.umi.count <- table(rdt.sub$inferred_umi)

    for (j in 1:length(unique.umi.count)) { 
      rdt.sub.sub <- rdt.sub[inferred_umi == names(unique.umi.count)[j], ]
      rdt[position %in% i & inferred_umi == names(unique.umi.count)[j], 
          inferred_pos := get.position.with.most.reads(rdt.sub.sub)]
    }
  }
  return (rdt)
}


get.pcr.duplicates <- function(rdt) {
  reads <- rdt[,.(inferred_umi, inferred_pos, rname)]
  unique.fragments <- unique(reads[order(reads$inferred_pos)])
  for (i in 1:nrow(unique.fragments)) {
    frag <- unique.fragments[i,]
    n <- nrow(rdt[inferred_umi==frag[[1]] & inferred_pos==frag[[2]],])
    unique.fragments[i, num := n]
  }
  return (unique.fragments)
}


plot.num.frag.per.umi <- function(umi.table, title) {
  dt <- data.table(umi.table)
  
  breaks <- pretty(range(dt$N), n = nclass.scott(dt$N), min.n = 1)
  bwidth <- diff(breaks)[1]
  
  plot(ggplot(dt, aes(N)) + 
    geom_histogram(aes(y=..count../sum(..count..)),
                   binwidth=bwidth, 
                   closed="left", boundary=0, fill="white", col="black") +
    theme_Publication() +
    ggtitle(title) +
    xlab("Number of reads per UMI") + ylab("UMI fraction"))
}


plot.base.fraction <- function(res.table, umi.length) {
  # Base percentage of UMIs
  base.matrix <- c()
  for(i in 1:umi.length) {
    # consider all fragments

    base.matrix <- rbindlist(list(base.matrix,
                                  data.table(t(data.frame(table(substr(unique(res.table[,umi]),
                                                                       i, i)), row.names=1)))),
                             use.names=T, fill=F, idcol=F)
  }
  
  #base.matrix.frac = sweep(base.matrix, 2, length(unique(umi.matrix[,umi])), "/")
  base.matrix.frac <- base.matrix / length(unique(res.table[,umi]))
  base.matrix.frac$ind <- 1:nrow(base.matrix.frac)
  
  base.matrix.frac.melt <- melt(base.matrix.frac, id.vars="ind")
  
  plot(ggplot(base.matrix.frac.melt, aes(ind, value, fill=variable)) +
         geom_bar(stat = "identity") + theme_Publication() + scale_fill_Publication() +
         theme(legend.title=element_blank()) + ggtitle("Original UMIs") +
         xlab("UMI position") + ylab("Percent"))
  
  
  base.inferred.matrix <- c()
  for(i in 1:umi.length) {
    base.inferred.matrix <- 
      rbindlist(list(base.inferred.matrix, 
                     data.table(t(data.frame(table(substr(unique(res.table[,inferred_umi]),
                                                          i, i)), row.names=1)))),
                use.names=T, fill=F, idcol=F)
  }
  
  base.inferred.matrix.frac <- base.inferred.matrix / length(unique(res.table[,inferred_umi]))
  
  base.inferred.matrix.frac$ind <- 1:nrow(base.inferred.matrix.frac)
  
  base.inferred.matrix.frac.melt <- melt(base.inferred.matrix.frac, id.vars="ind")
  
  plot(ggplot(base.inferred.matrix.frac.melt, aes(ind, value, fill=variable)) +
    geom_bar(stat = "identity") + theme_Publication() + scale_fill_Publication() +
    theme(legend.title=element_blank()) + ggtitle("Inferred UMIs") +
    xlab("UMI position") + ylab("Percent"))
}


plot.num.products.per.fragment <- function(pcr.products.dt) {
  # histogram of # products per fragment
  breaks <- pretty(range(pcr.products.dt$num),
                   n = nclass.scott(pcr.products.dt$num), min.n = 1)
  bwidth <- diff(breaks)[1]
  
  plot(ggplot(pcr.products.dt, aes(num)) +
         geom_histogram(aes(y=..count../sum(..count..)),
                        binwidth=bwidth, closed="left", boundary=0,
                        fill="white", col="black") + theme_Publication() +
    ggtitle("Number of PCR products per IVT fragment") +
    xlab("Number of PCR products") + ylab("Fraction"))
}


plot.stats.sam <- function(res.table, num.pcr.products.table, umi.length) {
  # calculate fragment number
  ori.fragments <- unique(res.table[,.(umi, position)])
  inf.fragments <- unique(res.table[,.(inferred_umi, inferred_pos)])
  
  # Calculate relative abundance of UMIs
  umi.table <- table(ori.fragments$umi)
  umi.inferred.table <- table(inf.fragments$inferred_umi)
  
  plot.num.frag.per.umi(umi.table, "Original UMIs")
  
  plot.num.frag.per.umi(umi.inferred.table, "Inferred UMIs")
  
  plot.base.fraction(res.table, umi.length)
  
  plot.num.products.per.fragment(num.pcr.products.table)
}

```








```{r}

# read in SAM file
samdt <- read.sam(test.sam)

# get reference sequence names
chr <- mixedsort(setdiff(unique(samdt[,rname]), "*"))
res.table <- list()
num.pcr.products <- list()


# for each reference sequence name (chr)
for (current.ref in chr) {
  # correct umi mismatch
  rdt <- umi.mismatch.correction(samdt, current.ref, umi.max.gap, umi.edit)
  # correct alignment position error
  rdt <- alignment.position.correction(rdt, pos.max.gap)
  
  
  # Now that UMIs have been fixed, go through each position and pick one read to
	# represent each UMI. Positions with more reads with that UMI and that are more 5'
	# are given higher priority. One read randomly selected to be the non-duplicate
  
  # Metrics to report:
  # Number of PCR priducts per IVT fragment
  num.pcr.products.table <- rbindlist(list(num.pcr.products.table, get.pcr.duplicates(rdt)),
                                use.names=F, fill=F, idcol=F)
  
  # Distribution of average UMI edit distance of all reads


	res.table <- rbindlist(list(res.table, 
	                   rdt[,c("rname", "position","inferred_pos",
	                          "umi", "inferred_umi")]),
	                   use.names=F, fill=F, idcol=F)
}

# Number of reads with mismatches in UMI
num.umi.mismatch <- nrow(res.table[umi != inferred_umi,])

# Number of reads with shifts in alignment position
num.pos.shift <- nrow(res.table[position != inferred_pos,])


stats.label <- c("Number of aligned reads:", 
                "Number of unique IVT fragments:", 
                "Percent unique IVT fragments:",
                "Number of unique UMIs:",
                "Number of unique UMIs after correction:",
                "Number of reads with mismatches in UMI:",
                "Percentage of reads with mismatches in UMI:",
                "Number of reads with shifts in alignment position:",
                "Percentage of reads with shifts in alignment position:",
                "Average number of products per fragment:",
                "Median number of products per fragment:")

stats <- c(nrow(res.table), 
          nrow(num.pcr.products.table),
          nrow(num.pcr.products.table)/nrow(res.table),
          length(unique(res.table$umi)), 
          length(unique(res.table$inferred_umi)),
          num.umi.mismatch,
          num.umi.mismatch/nrow(res.table),
          num.pos.shift,
          num.pos.shift/nrow(res.table),
          mean(num.pcr.products.table$num),
          median(num.pcr.products.table$num))

fwrite(data.table(stats.label, stats), umi.stats, sep="\t",
            quote=FALSE, row.names=FALSE, col.names=FALSE)

pdf(umi.qc)
plot.stats.sam(res.table, num.pcr.products.table, umi.length)
dev.off()
```


